jobs/taar_dynamo.py

```diff
@@ -0,0 +1,499 @@
+    jdata["start_date"] = start_date
+
+    # Filter out keys with an empty value
+    jdata = {key: value for key, value in list(jdata.items()) if value}
+
+    # We need to return a 4-tuple of values
+    # (numrec_dynamodb_pushed, json_list_length, json_list, error_json)
+
+    # These 4-tuples can be reduced in a map/reduce
+    return (0, 1, [jdata], [])
+
+
+def _write_batch(table, item_list, pause_time):
+    logging.info("Pausing for {} seconds".format(pause_time))
+    time.sleep(pause_time)
+    with table.batch_writer(overwrite_by_pkeys=["client_id"]) as batch:
+        for item in item_list:
+            batch.put_item(Item=item)
+
+
+class DynamoReducer(object):
+    def __init__(
+        self,
+        region_name=None,
+        table_name=None,
+        aws_access_key_id=None,
+        aws_secret_access_key=None,
+    ):
+
+        if region_name is None:
+            region_name = "us-west-2"
+
+        if table_name is None:
+            table_name = "taar_addon_data"
+
+        self._region_name = region_name
+        self._table_name = table_name
+
+        self._aws_access_key_id = aws_access_key_id
+        self._aws_secret_access_key = aws_secret_access_key
+
+    def hash_client_ids(self, data_tuple):
+        """
+        # Clobber the client_id by using sha256 hashes encoded as hex
+        # Based on the js code in Fx
+
+        """
+
+        for item in data_tuple[2]:
+            client_id = item["client_id"]
+            item["client_id"] = hash_telemetry_id(client_id)
+
+    def push_to_dynamo(self, data_tuple):  # noqa
+        """
+        This connects to DynamoDB and pushes records in `item_list` into
+        a table.
+
+        We accumulate a list of up to 50 elements long to allow debugging
+        of write errors.
+        """
+        # Transform the data into something that DynamoDB will always
+        # accept
+        # Set TTL to 60 days from now
+        ttl = int(time.time()) + 60 * 60 * 24 * 60
+
+        self.hash_client_ids(data_tuple)
+
+        item_list = [
+            {
+                "client_id": item["client_id"],
+                "TTL": ttl,
+                "json_payload": DynamoBinary(
+                    zlib.compress(json.dumps(item, default=json_serial).encode("utf8"))
+                ),
+            }
+            for item in data_tuple[2]
+        ]
+
+        # Clobber the enviroment variables for boto3 just prior to
+        # making the dynamodb connection
+        os.environ["AWS_ACCESS_KEY_ID"] = self._aws_access_key_id
+        os.environ["AWS_SECRET_ACCESS_KEY"] = self._aws_secret_access_key
+        # Obtain credentials from the singleton
+        conn = boto3.resource(
+            "dynamodb", region_name=self._region_name, config=BOTO_CFG
+        )
+
+        table = conn.Table(self._table_name)
+
+        retry_exceptions = (
+            "ProvisionedThroughputExceededException",
+            "ThrottlingException",
+        )
+
+        retries = 0
```

Comment:
@robotblake The `_write_batch` function was extracted so that it's more obvious that there is exponential backoff added around the boto3 write operation.